Fix rollout history bug

brianpursley opened the pull request:
#### What type of PR is this?
/kind bug
/kind cleanup

#### What this PR does / why we need it:
Fixes bug in kubectl rollout history where only the latest revision was displayed when a specific revision was requested and an output format was specified

Adds unit and integration tests for rollout history.

#### Which issue(s) this PR fixes:
Fixes #110097 

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Fixed bug in kubectl rollout history where only the latest revision was displayed when a specific revision was requested and an output format was specified
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

brianpursley commented:
Holding for review...
/hold

brianpursley commented:
/retest

ardaguclu commented on staging/src/k8s.io/kubectl/pkg/polymorphichelpers/history.go:
```diff
@@ -165,6 +157,25 @@ func (h *DeploymentHistoryViewer) ViewHistory(namespace, name string, revision i
+	for _, rs := range allRSs {
+		v, err := deploymentutil.Revision(rs)
+		if err != nil {
+			continue
```
Why we continue in here?. As far as I understand, error in here is not a fatal thing?

brianpursley replied:
I think I agree that it should return an error.  I was following the way that ViewHistory did it (see line 114 above), but it does seem that it would be misleading if it did not actually fail if there was an error.

Let me see about changing this to return an error...

brianpursley replied:
I think I'm going to leave this as continue for now, and have added warning messages if it skips a replicaset.

It looks like this was done intentionally in case it is unable to parse the revision from the replicaset's annotation, so that you never get into a situation where you can't see the rollout history because of a single corrupt replicaset's metadata.

I could be convinced that returning an error is the correct thing to do, it just gives me pause to change this since this is how it was already being done.  I wish there was a comment here or something explaining why continue is being used.  At least there will be a warning now.

ardaguclu commented on staging/src/k8s.io/kubectl/pkg/polymorphichelpers/history.go:
```diff
@@ -192,6 +203,25 @@ func (h *DaemonSetHistoryViewer) ViewHistory(namespace, name string, revision in
+	for _, h := range history {
+		applied, err := applyDaemonSetHistory(ds, h)
+		if err != nil {
+			return nil, err
```
For example, we are returning err here instead continue

ardaguclu commented on staging/src/k8s.io/kubectl/pkg/polymorphichelpers/history.go:
```diff
@@ -259,6 +289,43 @@ func (h *StatefulSetHistoryViewer) ViewHistory(namespace, name string, revision
+	if err != nil {
+		return nil, fmt.Errorf("failed to retrieve replica sets from deployment %s: %v", name, err)
+	}
+	allRSs := oldRSs
```
Why we introduce `allRSs` instead using `oldRSs`?

brianpursley replied:
I just extracted the original code into a func, but didn't try to improve it.  I think you're right.  It is unnecessary to introduce this extra variable

ardaguclu commented on test/cmd/apps.sh:
```diff
@@ -440,6 +440,40 @@ run_deployment_tests() {
   kubectl delete configmap test-set-env-config "${kube_flags[@]:?}"
   kubectl delete secret test-set-env-secret "${kube_flags[@]:?}"
 
+  ### Get rollout history
```
I guess, it is difficult to add test cases also for statefulset and daemonsets because of their nature?

brianpursley replied:
It shouldn't be.  Let me see if I can add some

ardaguclu commented on staging/src/k8s.io/kubectl/pkg/polymorphichelpers/history.go:
```diff
@@ -48,6 +48,7 @@ const (
 // HistoryViewer provides an interface for resources have historical information.
 type HistoryViewer interface {
 	ViewHistory(namespace, name string, revision int64) (string, error)
+	GetHistory(namespace, name string) (map[int64]runtime.Object, error)
```
I understand that we can not easily remove `ViewHistory`. But just curiosity; semantically, GetHistory also **contains** `ViewHistory`?.

ardaguclu replied:
I mean consumers can do the same behavior just using the `GetHistory` without needing `ViewHistory`

brianpursley replied:
GetHistory is just to get the data associated with the history, while ViewHistory gets *and* prints the history.

You are correct, a consumer can use GetHistory alone if they want to control how the data is handled.  Or they can use ViewHistory alone if they want to let it get and print the output in the predefined way.

---

I thought about trying to refactor ViewHistory, but decided it would be too big of a change for this PR.  I do think it probably needs to be refactored though.  Maybe as a follow-up task.

Ideally, I think it would be good for ViewHistory to only be concerned with printing the history, and that it should not need to know how to also retrieve the history data.  Maybe even move it to a describer like some of the existing TODO comments suggest.

ardaguclu commented on staging/src/k8s.io/kubectl/pkg/polymorphichelpers/history.go:
```diff
@@ -101,24 +102,16 @@ type DeploymentHistoryViewer struct {
 	for _, rs := range allRSs {
 		v, err := deploymentutil.Revision(rs)
 		if err != nil {
+			klog.Warningf("unable to get revision from replicaset %s for deployment %s in namespace %s", rs.Name, name, namespace)
```
nit: would it be better to lower log level(ie. `klog.v(2)`) and embed error string into the message?

brianpursley replied:
Thanks.  I added the error to the message.

But I'm leaving it as a warning, for now at least.

I don't think there would be a normal situation where an error occurs while getting the replicaset revision, so my intention is that the caller be made aware of the problem by printing a warning message.  If this happens, then either there is something corrupt about that replicaset, or perhaps a bug in the revision logic somewhere in k8s that causes that situation (should it even occur).

ardaguclu replied:
> I don't think there would be a normal situation where an error occurs while getting the replicaset revision, so my intention is that the caller be made aware of the problem by printing a warning message. If this happens, then either there is something corrupt about that replicaset, or perhaps a bug in the revision logic somewhere in k8s that causes that situation (should it even occur).

Good point

ardaguclu commented:
This lgtm but I'll defer it other folks to review.

brianpursley committed:
Fix rollout history bug

Fix rollout history bug where the latest revision was
always shown when requesting a specific revision and
specifying an output.

Add unit and integration tests for rollout history.

brianpursley commented:
ping @KnVerey thoughts on this PR?

ardaguclu commented:
/lgtm

dims commented:
@brianpursley need to remove hold for this to land

brianpursley commented:
/unhold
